<a href="https://colab.research.google.com/github/VeronikaBastiuchenko/Pandas-Numpy/blob/main/hw_3_3_%D0%A0%D0%BE%D0%B7%D1%88%D0%B8%D1%80%D0%B5%D0%BD%D1%96_%D0%BC%D0%B5%D1%82%D0%BE%D0%B4%D0%B8_%D0%BE%D0%B1%D1%80%D0%BE%D0%B1%D0%BA%D0%B8_%D0%B4%D0%B0%D0%BD%D0%B8%D1%85_%D1%83_Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Домашнє завдання до теми "Розширені методи обробки даних у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [3]:
import pandas as pd
df = pd.read_csv('supermarket_sales.csv')

1. Відсортуйте дані за спаданням ціни товара ('Unit price') і виведіть 5 перших рядочків на екран.

In [4]:
sorted_df = df.sort_values(by='Unit price', ascending=False)
print("Топ-5 за ціною товару:")
print(sorted_df.head(5))

Топ-5 за ціною товару:
      Invoice ID Branch       City Customer type Gender         Product line  \
983  148-41-7930      C  Naypyitaw        Normal   Male    Health and beauty   
122  219-22-9386      B   Mandalay        Member   Male    Sports and travel   
930  641-62-7288      B   Mandalay        Normal   Male   Home and lifestyle   
494  437-53-3084      B   Mandalay        Normal   Male  Fashion accessories   
283  667-92-0055      A     Yangon        Member   Male    Health and beauty   

     Unit price  Quantity  Tax 5%    Total       Date   Time      Payment  \
983       99.96         7  34.986  734.706  1/23/2019  10:33         Cash   
122       99.96         9  44.982  944.622   3/9/2019  17:26  Credit card   
930       99.92         6  29.976  629.496  3/24/2019  13:33      Ewallet   
494       99.89         2   9.989  209.769  2/26/2019  11:48      Ewallet   
283       99.83         6  29.949  628.929   3/4/2019  15:02      Ewallet   

       cogs  gross margin percent

2. Знайдіть інвойс (рядок в даних) в категорії товарів `Electronic accessories`, в якому був куплений товар з найбільшою кількістю і найбільшою ціною. Виведіть цей рядок даних на екран.

In [5]:
ea_df = df[df['Product line'] == 'Electronic accessories']
max_row = ea_df.loc[ea_df[['Unit price', 'Quantity']].product(axis=1).idxmax()]
print("\nІнвойс з найбільшою кількістю і ціною в Electronic accessories:")
print(max_row)


Інвойс з найбільшою кількістю і ціною в Electronic accessories:
Invoice ID                            817-69-8206
Branch                                          B
City                                     Mandalay
Customer type                              Normal
Gender                                     Female
Product line               Electronic accessories
Unit price                                  99.73
Quantity                                        9
Tax 5%                                    44.8785
Total                                    942.4485
Date                                     3/2/2019
Time                                        19:42
Payment                               Credit card
cogs                                       897.57
gross margin percentage                  4.761905
gross income                              44.8785
Rating                                        6.5
Name: 209, dtype: object


3.1. Створіть нову колонку в даних `Rating Category`, яка визначає категорію рейтингу (`Rating`): високий, середній чи низький за наступним правилом:
- якщо рейтинг вищий або рівний 8 - то він високий (`High`)
- якщо рейтинг від 6 до 8 не включаючи - то він середній (`Medium`)
- інакше - низький (`Low`).

In [7]:
def categorize_rating(r):
    if r >= 8:
        return 'High'
    elif r >= 6:
        return 'Medium'
    else:
        return 'Low'
df['Rating Category'] = df['Rating'].apply(categorize_rating)

3.2. Виведіть, який відсоток даних становить кожне зі значень категорій рейтингу.

In [16]:
rating_percent = df['Rating Category'].value_counts(normalize=True) * 100
print("Відсотковий розподіл категорій рейтингу:")
print(rating_percent)

Відсотковий розподіл категорій рейтингу:
Rating Category
Medium    35.6
High      32.9
Low       31.5
Name: proportion, dtype: float64


4.1. Запустіть код нижче аби створити змінну `time`. Ця змінна містить час в форматі "година:хвилина". Напишіть функцію `get_hour`, яка буде зі змінної формату як `time` дістати лише години і конвертувати їх у тип `int`.

Запустіть написану функцію на змінній `time` та перевірте правильність роботи функції.

In [12]:
time = '11:28'

In [15]:
def get_hour(t):
    return int(t.split(':')[0])

print("Година з time =", time, ":", get_hour(time))

Година з time = 11:28 : 11


4.2. З допомогою функції `get_hour` і метода `DataFrame.apply()` створіть нову колонку `Hour` в нашому наборі даних, яка буде містити години здійснення покупки, обчислені на основі часу здійснення покупки `Time`.

In [17]:
df['Hour'] = df['Time'].apply(get_hour)

4.3. Виведіть кількість значень за кожною годиною покупки і дайте відповідь "О яких годинах найбільше здійснюють покупок покупці і чому на Ваш погляд саме о цих годинах? Чи це слогічно, що найбільше покупок саме в ці години?".

In [23]:
hour_counts = df['Hour'].value_counts().sort_index()
print("Кількість покупок за кожною годиною:")
print(hour_counts)

Кількість покупок за кожною годиною:
Hour
10    101
11     90
12     89
13    103
14     83
15    102
16     77
17     74
18     93
19    113
20     75
Name: count, dtype: int64


Виходячи з наданих результатів, найбільше покупок здійснюється о 19:00 (113 покупок), за якою слідують 13:00 (103 покупки), 15:00 (102 покупки), та 10:00 (101 покупка). Це цілком логічно, оскільки ці години часто припадають на обідню перерву або вечірній час після закінчення робочого дня, коли люди мають більше вільного часу для шопінгу.

5. Виведіть середнє значення чеку (`Total`) для покупок в містах, наявних в даних, за допомогою функції `groupby`.

In [22]:
mean_total_by_city = df.groupby('City')['Total'].mean()
print("Середній чек по містах:")
print(mean_total_by_city)

Середній чек по містах:
City
Mandalay     319.872506
Naypyitaw    337.099715
Yangon       312.354031
Name: Total, dtype: float64


6. Виведіть кількість покупок, середнє, мінімальне та максимальне значення для загальної суми чеку (Total) в містах, наявних в даних, за допомогою функції groupby.
Зробіть висновки з цих показників та дайте рекомендації для мережі супермаркетів.

In [21]:
summary_stats = df.groupby('City')['Total'].agg(['count', 'mean', 'min', 'max'])
print("Статистика Total по містах:")
print(summary_stats)

Статистика Total по містах:
           count        mean      min      max
City                                          
Mandalay     332  319.872506  18.6375  1022.49
Naypyitaw    328  337.099715  10.6785  1042.65
Yangon       340  312.354031  12.6945  1039.29


Naypyitaw має найвищий середній чек ($337.10) та найбільш максимальний чек, що свідчить про потенціал для більших витрат з боку покупців. Mandalay та Yangon мають схожі середні та максимальні показники, хоча Yangon має найбільшу кількість покупок (340). Рекомендації для супермаркетів:Для Naypyitaw краще зосередитись на преміальних товарах та програмах лояльності для стимулювання більших чеків. У Yangon можна оптимізувати швидкість обслуговування та наявність товарів через велику кількість покупок. Для Mandalay можна впровадити акції чи інші способи щоб збільшити середній чек.

7. Ми внесли деякі зміни в наш набір даних, тож, давайте запишемо результат у файл. Запишіть оновлену версію даних у файл `supermarket_sales_enriched.csv`. Зверніть увагу, аби не записати нічого зайвого, наприклад, індекси рядків нам в файлі не потрібні, оскільки вони стандартні.

In [26]:
df.to_csv('supermarket_sales_enriched.csv', index=False)